### DISCUSSION TITLE
5xx errors when `enable_buffering` was enabled in the plugin

### DISCUSSION THREAD

**Discussion by shettyh** [Kong/kong#9304] [Help]
Is there restriction on for what kind of requests we should `enable_buffering` for ? 

Error: 
```
stack traceback:
coroutine 0:
[C]: in function 'error'
/usr/local/share/lua/5.1/kong/pdk/private/phases.lua:84: in function 'check_phase'
/usr/local/share/lua/5.1/kong/pdk/request.lua:472: in function 'get_header'
/usr/local/share/lua/5.1/kong/error_handlers.lua:68: in function 'access'
access_by_lua(nginx-kong.conf:101):2: in main chunk while sending to client

lua entry thread aborted: runtime error: /usr/local/share/lua/5.1/kong/pdk/private/phases.lua:84: function cannot be called in balancer phase (only in: rewrite, access, cluster_listener, header_filter, body_filter, log)
upstream prematurely closed connection while reading response header from upstream, client: 203.199.108.150, server: kong, request: \"POST /v1/fund_accounts/validations HTTP/1.1\", subrequest: \"/kong_buffered_http\
```

Kong Version: 2.0.3

**Comment by samugi** [comment]
Hello @shettyh, are you perhaps attempting to call `enable_buffering()` from the `balancer` phase of a custom plugin?
If so, that is not allowed. Can you try enabling buffering from the `access` or `rewrite` phase instead?

**Reply by shettyh** [reply]
@samugi no we are enabling it in access phase only. Not failing for all the requests though

**Comment by flrgh** [comment]
:wave: hi @shettyh

It looks like this is a bug that was fixed in the 2.3.1 release:

https://github.com/Kong/kong/blob/master/CHANGELOG.md#231
https://github.com/Kong/kong/pull/6735

I would recommend updating to the newest release ([2.8.1](https://github.com/Kong/kong/releases/tag/2.8.1) as of this writing) as it contains the fix for this and many, many other improvements.

**Reply by shettyh** [reply]
@flrgh thanks for the response. Yes we will be upgrading to latest version soon. When the `enable_buffering` was enabled we saw an increased rate of 5xx ( mostly 502 and 500 ), not all 500 have this error in the log. Can any other reason for this ? Is there any restriction on which type of request we should enable buffering, like HEAD, GET requests should ignored etc ??

**Reply by shettyh** [reply]
@bungle if you can help here please

**Reply by samugi** [reply]
@shettyh you said "not all 500 have this error in the log": are you seeing other errors in Kong? I am asking because it's possible the `500` response is generated by the upstream API. I would also recommend upgrading to the newest release, then checking if the behaviour persists, in which case then it is important to differentiate errors originating in Kong from those returned by the backend API. One thing to keep in mind is that buffered proxying cannot be enabled over http/2.

**Reply by shettyh** [reply]
- All the errors 500 and 502s are returned from kong `( source = exit )`
- All the requests are http/1.1

@samugi ^

**Reply by flrgh** [reply]
I can't think of any reason why `enable_buffering` would _increase_ the rate of 502 errors--these are generally the result of a TCP error or an invalid HTTP response received from the upstream. The bug I referenced could potentially cause some 502 errors to be reported as 500 due to the unhandled error though.

**Reply by shettyh** [reply]
@flrgh thanks for the response. Other than above mentioned stack trace we are not seeing any other error with `debug` log enabled, might need to fallback to `body_filter` based body reading

**Reply by samugi** [reply]
@shettyh please let us know if you encounter the problem after upgrading: as you mentioned the only error detected is the one that should be taken care of by the more recent releases.

**Reply by shettyh** [reply]
Dont see any other error in the log, so dont know what are the other errors 😢 .
